Port redirect off of XML/XSLT

philipcaisip opened the pull request:
Removes client-side redirects off of the XSLT registration scripts. Uses HTML custom elements, like #1837.

Currently it seems to be functionally correct but the code does define the redirect tag in two different spots ([one in WrongStepAjaxInterceptor](https://github.com/BorderTech/wcomponents/blob/c682a6cf75f9840adcc2eeb9e85824a20874524b/wcomponents-core/src/main/java/com/github/bordertech/wcomponents/container/WrongStepAjaxInterceptor.java#L39C1-L42) and [one in ContentRenderer](https://github.com/BorderTech/wcomponents/blob/c682a6cf75f9840adcc2eeb9e85824a20874524b/wcomponents-core/src/main/java/com/github/bordertech/wcomponents/render/webxml/WContentRenderer.java#L19-L22)). This likely ought to be in one spot but I wasn't sure which one spot.

philcaisip committed:
Add redirect unit tests and update location change

Unit tests use a mock of window.parent, since that's the object that
the implementation uses (not window directly). Note, however, that in
real cases the parent just points back to the same window itself
anyways.

The window location/url switch has also been changed. The method
.replace() is now used instead of setting href directly, as advised in
https://developer.mozilla.org/en-US/docs/Web/API/Location/href. This
stops the original page from being saved in session history, making it
resemble actual redirection instead of just navigation.

philcaisip committed:
Add redirect unit tests and update location change

Unit tests use a mock of window.parent, since that's the object that
the implementation uses (not window directly). Note, however, that in
real cases the parent just points back to the same window itself
anyways.

The window location/url switch has also been changed. The method
.replace() is now used instead of setting href directly, as advised in
https://developer.mozilla.org/en-US/docs/Web/API/Location/href. This
stops the original page from being saved in session history, making it
resemble actual redirection instead of just navigation.

philcaisip committed:
Merge branch 'feature/redirect' of https://github.com/BorderTech/wcomponents into feature/redirect

philcaisip committed:
Add custom element code to replace redirect XSLT

There's currently a failing WContentRenderer test that still needs to be
taken care of. It seems to fail whether the custom html elements are
used or the original XSLT. Will need to investigate.

philcaisip committed:
Update redirect unit tests and remove XSLT

Registration code for redirects now completely deleted from all.xsl (the
template is still there, just not used). Problem from last commit was
fixed (it was a namespace issue) and the remaining WrongStepAjax-
Interceptor tests were also updated.

philcaisip committed:
Add missing Javadoc comments

philcaisip committed:
Merge branch 'hermione' into feature/redirect

ricksbrown commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/container/WrongStepAjaxInterceptor.java:
```diff
@@ -170,7 +175,7 @@ private void handleRenderRedirect(final PrintWriter writer) {
 
 		// Redirect URL
-		writer.print("<ui:redirect url=\"" + redirectUrl + "\" />");
+		writer.print(String.format("<%s url=\"%s\"></%s>", TAG_REDIRECT, redirectUrl, TAG_REDIRECT));
```
We don't repeat args to `String.format` and similar interpolation methods. There is always a way to reuse the argument again in a different place. I'll let you work out how 😉

ricksbrown replied:
Ok that works but can you do it while still using `String.format` and without repeating the args?

Clue: read about [argument_index](https://docs.oracle.com/javase/8/docs/api/java/util/Formatter.html#syntax)

ricksbrown commented on wcomponents-theme/src/test/spec/wc.ui.redirect.test.mjs:
```diff
@@ -0,0 +1,81 @@
+import redirect from "wc/ui/redirect.mjs";
+
+describe("wc/ui/redirect", () => {
```
You have at least one unused variable. I'll let you find it...
You are also seeing red errors in IntelliJ from TSC... Don't ignore them.

ricksbrown commented on wcomponents-theme/src/test/spec/wc.ui.redirect.test.mjs:
```diff
@@ -0,0 +1,81 @@
+	});
+
+	it("recognises attachment links as launch links", () => {
+		spyOn(redirect, "isLaunchUrl").and.callThrough();
```
I think you may be overusing the spy here (and elsewhere in this suite)? What value does it add?
For example, this test explicitly calls `redirect.isLaunchUrl` and then uses a spy to verify that this call happened.
Could this not be simplified to something like this?

```js
	it("recognises attachment links as launch links", () => {
		expect(redirect.isLaunchUrl("mailto:test@email.com")).toBeTrue();
	});
```

ricksbrown commented on wcomponents-theme/src/test/spec/wc.ui.redirect.test.mjs:
```diff
@@ -0,0 +1,81 @@
+
+	beforeEach(() => {
+		window.parent = {
+			location : {
```
no space before `:` except for the ternary operator

For example this line should be:

`location: {`

ricksbrown commented on wcomponents-theme/src/test/spec/wc.ui.redirect.test.mjs:
```diff
@@ -0,0 +1,81 @@
+	beforeEach(() => {
+		window.parent = {
+			location : {
+				storedUrl : "",
```
Regarding `storedUrl` fair solution. I would personally not do it this way mainly because `window.location` actually has a [defined interface](https://html.spec.whatwg.org/multipage/nav-history-apis.html#the-location-interface) and the property `storedUrl` is not part of it right?

Functionally no problem but some of the tools will get angry.
I would simply move `storedUrl` off the object and into the `beforeEach` function as a variable (with `let` not `const`).

philcaisip committed:
Clean up redirect unit tests

Simplifies (i.e. gets rid of unnecessary spies in Jasmine tests) and
cleans up some code, as suggested in PR #1841.

philcaisip committed:
Merge branch 'hermione' into feature/redirect

philipcaisip commented:
Hi @ricksbrown, I've made those changes that you mentioned. Regarding the red TSC errors though, I couldn't see any. There are a few yellow warnings but they look unavoidable/minor (just complaining about some weird Jasmine syntax and the mock window). Is this fine?

ricksbrown commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/container/WrongStepAjaxInterceptor.java:
```diff
@@ -170,7 +175,7 @@ private void handleRenderRedirect(final PrintWriter writer) {
 
 		// Redirect URL
-		writer.print("<ui:redirect url=\"" + redirectUrl + "\" />");
+		writer.print(String.format("<%1$s url=\"%2$s\"></%1$s>", TAG_REDIRECT, redirectUrl));
```
Nice! 
No need to overuse the argument index, you should be able to only use it where you need it, so it could look something like this:

```java
writer.print(String.format("<%s url=\"%s\"></%1$s>", TAG_REDIRECT, redirectUrl));
```

ricksbrown commented:
> Hi @ricksbrown, I've made those changes that you mentioned. Regarding the red TSC errors though, I couldn't see any. There are a few yellow warnings but they look unavoidable/minor (just complaining about some weird Jasmine syntax and the mock window). Is this fine?

Yes, it's more a general point. They probably got sorted out when you cleaned up the code a bit.

philcaisip committed:
Clean up String.format call

Gets rid of unnecessary argument indices, as suggested in code review.

ricksbrown commented:
LGTM - you can merge it when you are ready.
